### PR TITLE
[cu-23v2xkb][Pablo] ConstantProductPool Integration

### PR DESCRIPTION
**Pull request by vimukthi-git**
- Integrate constant product pool dex(Uniswap) in to pallet-pablo

### CODE REVIEW

File: frame/pablo/src/uniswap.rs [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+#[cfg(test)]
+mod tests {
+	use crate::mock::{new_test_ext, ALICE, BOB, BTC, USDT};
+	use composable_traits::{defi::CurrencyPair, dex::Amm};
+	use frame_support::assert_ok;
+	use sp_arithmetic::Permill;
+
+	#[test]
+	fn test() {
+		new_test_ext().execute_with(|| {});
```
Comment (vimukthi-git) [inline frame/pablo/src/uniswap.rs:338]: TODO

File: frame/pablo/src/lib.rs [hunk 2]
```diff
@@ -56,13 +60,14 @@ pub mod pallet {
...
 		StableSwap(StableSwapPoolInfo<AccountId, AssetId>),
+		ConstantProduct(ConstantProductPoolInfo<AccountId, AssetId>),
 	}
 
 	type AssetIdOf<T> = <T as Config>::AssetId;
 	type BalanceOf<T> = <T as Config>::Balance;
 	type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
 	type PoolIdOf<T> = <T as Config>::PoolId;
-	type PoolOf<T> =
+	pub type PoolConfigurationOf<T> =
```
Comment (vimukthi-git) [inline frame/pablo/src/lib.rs:70]: Minor rename @vivekvpandya

File: frame/pablo/src/lib.rs [hunk 3]
```diff
@@ -324,10 +360,19 @@ pub mod pallet {
 					Self::deposit_event(Event::PoolCreated { pool_id, owner: pool.owner.clone() });
 					Ok(pool_id)
 				},
+				PoolConfiguration::ConstantProduct(constant_product_pool_info) =>
+					Uniswap::<T>::do_create_pool(
```
Comment (vimukthi-git) [inline frame/pablo/src/lib.rs:364]: Uniswap module can be called in the relevant places including the Pallet AMM impl functions

**Comment by vimukthi-git** [comment]
What's left: Need to fix some tests still

File: frame/pablo/src/lib.rs [hunk 4]
```diff
@@ -122,6 +130,28 @@ pub mod pallet {
 			/// Amount of minted lp tokens.
 			mint_amount: T::Balance,
 		},
+		/// Liquidity added into the pool `T::PoolId`.
```
Comment (vivekvpandya) [inline frame/pablo/src/lib.rs:133]: Do we need this change?

File: frame/pablo/src/lib.rs [hunk 5]
```diff
@@ -412,6 +450,10 @@ pub mod pallet {
 			let base_amount = Self::get_exchange_value(pool_id, pair.base, quote_amount)?;
 			let base_amount_u: u128 = T::Convert::convert(base_amount);
 
+			// https://uniswap.org/whitepaper.pdf
```
Comment (vivekvpandya) [inline frame/pablo/src/lib.rs:453]: I think reference to uniswa-paper is not required here. Please also note that if we want to compute fees differently for `Curve` and `Uniswap` we can specialize that part with a new function like `compute_fee()`. If current code is fine then `Uniswap-v2` 's test related to fees should fail and expected values will be required to be adjusted.

File: frame/pablo/src/lib.rs [hunk 6]
```diff
@@ -122,6 +130,28 @@ pub mod pallet {
...
+		LiquidityAddedToLiquidityBootstrappingPool {
+			/// Pool id to which liquidity added.
+			pool_id: T::PoolId,
+			/// Amount of base asset deposited.
+			base_amount: T::Balance,
+			/// Amount of quote asset deposited.
+			quote_amount: T::Balance,
+		},
+		/// Liquidity added into the pool `T::PoolId`.
+		LiquidityAddedToConstantProductPool {
```
Comment (vivekvpandya) [inline frame/pablo/src/lib.rs:143]: LiquidityAdded events for ConstantProduct and StableSwap are same. And LBP can also use that same event without any issue.

File: frame/pablo/src/lib.rs [hunk 7]
```diff
@@ -592,7 +678,53 @@ pub mod pallet {
...
+				ConstantProduct(constant_product_pool_info) => {
+					let (base_amount, quote_amount, lp_fees, owner_fees) = Self::do_compute_swap(
+						pool_id,
+						pair,
+						quote_amount,
+						true,
+						constant_product_pool_info.fee,
+						constant_product_pool_info.owner_fee,
+					)?;
+					let total_fees = lp_fees.safe_add(&owner_fees)?;
```
Comment (vivekvpandya) [inline frame/pablo/src/lib.rs:690]: this is contradicting to do_compute_swap(), as noted if we want we can specialize fee computation method.

File: frame/pablo/src/lib.rs [hunk 8]
```diff
@@ -75,21 +76,28 @@ pub mod pallet {
...
 		StableSwap(StableSwapPoolInfo<AccountId, AssetId>),
+		ConstantProduct(ConstantProductPoolInfo<AccountId, AssetId>),
 	}
 
 	type AssetIdOf<T> = <T as Config>::AssetId;
 	type BalanceOf<T> = <T as Config>::Balance;
 	type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
 	type PoolIdOf<T> = <T as Config>::PoolId;
-	type PoolConfigurationOf<T> =
+	pub type PoolConfigurationOf<T> =
```
Comment (vivekvpandya) [inline frame/pablo/src/lib.rs:96]: Is this `pub` required ?

Comment (vimukthi-git) [reply frame/pablo/src/lib.rs]: Good catch, it was from an earlier version

File: frame/pablo/src/lib.rs (hunk 6 above)
Comment (vimukthi-git) [reply frame/pablo/src/lib.rs:143]: Idea is to keep it this way till reach cu-23v2y3n

File: frame/pablo/src/lib.rs (hunk 5 above)
Comment (vimukthi-git) [reply frame/pablo/src/lib.rs]: I was using this while trying to understand this function if it's the same as what uniswap has, IMO we need to move this function to `stable_swap.rs` as the logic is specific to that.

File: frame/pablo/src/lib.rs (hunk 7 above)
Comment (vimukthi-git) [reply frame/pablo/src/lib.rs:686]: Yes, got a bit confused by `do_compute_swap` function in this file.

File: frame/pablo/src/lib.rs (hunk 5 above)
Comment (vimukthi-git) [reply frame/pablo/src/lib.rs]: fixed

File: frame/pablo/src/lib.rs (hunk 7 above)
Comment (vimukthi-git) [reply frame/pablo/src/lib.rs:686]: fixed

Comment (vivekvpandya) [reply frame/pablo/src/lib.rs:686]: so basically we want to specialize dp_compute_swap(). Okay then in next PR I will create it in stable_swap.

File: frame/pablo/src/lib.rs (hunk 6 above)
Comment (vivekvpandya) [reply frame/pablo/src/lib.rs:143]: This change does not bring in LBP to pablo. So perhaps we can remove `LiquidityAddedToLiquidityBootstrappingPool` and just have `LiquidityAdded` which is used for both StableSwap and Uniswap

**Review by vivekvpandya — DISMISSED** [review]

**Review by vivekvpandya — APPROVED** [review]

**Pull request merged**